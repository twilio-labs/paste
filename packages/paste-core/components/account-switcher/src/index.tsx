export * from "./AccountSwitcher";
export * from "./AccountSwitcherGroup";
export * from "./AccountSwitcherSeparator";
export * from "./AccountSwitcherItem";
export * from "./AccountSwitcherItemRadio";
export * from "./AccountSwitcherBadge";
export * from "./useAccountSwitcher";
