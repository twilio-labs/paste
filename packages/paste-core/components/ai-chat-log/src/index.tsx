export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageActionGroup } from "./AIChatMessageActionGroup";
export type { AIChatMessageActionGroupProps } from "./AIChatMessageActionGroup";
export { AIChatMessageBody } from "./AIChatMessageBody";
export type { AIChatMessageBodyProps } from "./AIChatMessageBody";
export { AIChatMessageAuthor } from "./AIChatMessageAuthor";
export type { AIChatMessageAuthorProps } from "./AIChatMessageAuthor";
export { AIChatMessageActionCard } from "./AIChatMessageActionCard";
export type { AIChatMessageActionCardProps } from "./AIChatMessageActionCard";
export { AIChatMessageLoading } from "./AIChatMessageLoading";
export type { AIChatMessageLoadingProps } from "./AIChatMessageLoading";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger, AIPartialIDChat, AIChat } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";

export type { AIMessageVariants } from "./AIMessageContext";
