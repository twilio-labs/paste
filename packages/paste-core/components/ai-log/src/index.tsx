export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageMeta } from "./AIChatMessageMeta";
export type { AIChatMessageMetaProps } from "./AIChatMessageMeta";
export { AIChatMessageMetaItem } from "./AIChatMessageMetaItem";
export type { AIChatMessageMetaItemProps } from "./AIChatMessageMetaItem";
export { AIChatMessageContent } from "./AIChatMessageContent";
export type { AIChatMessageContentProps } from "./AIChatMessageContent";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";
