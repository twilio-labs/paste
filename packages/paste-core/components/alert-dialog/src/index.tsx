import * as React from 'react';
import * as PropTypes from 'prop-types';
import {useUID} from '@twilio-paste/uid-library';
import {useTransition} from '@twilio-paste/animation-library';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';
import {ModalDialogOverlay} from '@twilio-paste/modal';
import {AlertDialogHeader} from './AlertDialogHeader';
import {AlertDialogBody} from './AlertDialogBody';
import {AlertDialogContent} from './AlertDialogContent';
import {AlertDialogFooter} from './AlertDialogFooter';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getAnimationStates = (): any => ({
  from: {opacity: 0, transform: `scale(0.675)`},
  enter: {opacity: 1, transform: `scale(1)`},
  leave: {opacity: 0, transform: `scale(0.675)`},
  // https://www.react-spring.io/docs/hooks/api
  config: {
    mass: 0.5,
    tension: 370,
    friction: 26,
  },
});

export interface AlertDialogProps extends React.HTMLAttributes<HTMLElement>, Pick<BoxProps, 'element'> {
  children: NonNullable<React.ReactNode>;
  destructive?: boolean;
  heading: string;
  isOpen: boolean;
  onConfirm: () => void;
  onConfirmLabel: string;
  onDismiss: () => void;
  onDismissLabel: string;
}

export const AlertDialog = React.forwardRef<HTMLDivElement, AlertDialogProps>(
  (
    {
      children,
      destructive,
      element = 'ALERT_DIALOG',
      heading,
      isOpen,
      onConfirm,
      onConfirmLabel,
      onDismiss,
      onDismissLabel,
      ...props
    },
    ref
  ) => {
    const transitions = useTransition(isOpen, getAnimationStates());
    const headingID = useUID();
    const bodyID = useUID();

    return (
      <>
        {transitions(
          (styles, item) =>
            item && (
              <ModalDialogOverlay isOpen={isOpen} style={{opacity: styles.opacity}}>
                <Box
                  // @ts-expect-error Render overlay as box for customization
                  as={AlertDialogContent}
                  {...safelySpreadBoxProps(props)}
                  aria-labelledby={headingID}
                  aria-describedby={bodyID}
                  element={element}
                  ref={ref}
                  role="alertdialog"
                  style={styles}
                >
                  <AlertDialogHeader headingID={headingID} element={`${element}_HEADER`}>
                    {heading}
                  </AlertDialogHeader>
                  <AlertDialogBody bodyID={bodyID} element={`${element}_BODY`}>
                    {children}
                  </AlertDialogBody>
                  <AlertDialogFooter
                    destructive={destructive}
                    element={`${element}_FOOTER`}
                    onDismiss={onDismiss}
                    onDismissLabel={onDismissLabel}
                    onConfirm={onConfirm}
                    onConfirmLabel={onConfirmLabel}
                  />
                </Box>
              </ModalDialogOverlay>
            )
        )}
      </>
    );
  }
);

AlertDialog.displayName = 'AlertDialog';

AlertDialog.propTypes = {
  children: PropTypes.node.isRequired,
  destructive: PropTypes.bool,
  element: PropTypes.string,
  heading: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onConfirm: PropTypes.func.isRequired,
  onConfirmLabel: PropTypes.string.isRequired,
  onDismiss: PropTypes.func.isRequired,
  onDismissLabel: PropTypes.string.isRequired,
};
