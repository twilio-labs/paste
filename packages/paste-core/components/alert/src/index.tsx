import * as React from 'react';
import * as PropTypes from 'prop-types';
import {ValueOf} from '@twilio-paste/types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import {MediaObject, MediaFigure, MediaBody} from '@twilio-paste/media-object';
import {Button} from '@twilio-paste/button';
import {CloseIcon} from '@twilio-paste/icons/esm/CloseIcon';
import {ErrorIcon} from '@twilio-paste/icons/esm/ErrorIcon';
import {InformationIcon} from '@twilio-paste/icons/esm/InformationIcon';
import {WarningIcon} from '@twilio-paste/icons/esm/WarningIcon';

export const AlertRoles = {
  ERROR: 'alert',
  NEUTRAL: 'status',
  WARNING: 'alert',
} as const;
export const AlertVariants = {
  ERROR: 'error',
  NEUTRAL: 'neutral',
  WARNING: 'warning',
} as const;
export const AlertBackgroundColors = {
  ERROR: 'colorBackgroundErrorLightest',
  NEUTRAL: 'colorBackgroundNeutralLightest',
  WARNING: 'colorBackgroundWarningLightest',
} as const;
export const AlertBorderColors = {
  ERROR: 'colorBorderErrorLight',
  NEUTRAL: 'colorBorderNeutralLight',
  WARNING: 'colorBorderWarningLight',
} as const;

export type AlertVariants = ValueOf<typeof AlertVariants>;
export type AlertBackgroundColors = ValueOf<typeof AlertBackgroundColors>;
export type AlertBorderColors = ValueOf<typeof AlertBorderColors>;
export type AlertRoles = ValueOf<typeof AlertRoles>;

export interface AlertProps {
  id?: never;
  className?: never;
  children: NonNullable<React.ReactNode>;
  onDismiss?: () => void;
  role?: string;
  variant: AlertVariants;
}

const renderAlertIcon = (variant: AlertVariants): React.ReactElement => {
  switch (variant) {
    case AlertVariants.ERROR:
      return <ErrorIcon iconColor="colorTextErrorDark" decorative={false} title="error: " size="sizeIcon30" />;
    case AlertVariants.WARNING:
      return <WarningIcon iconColor="colorTextWarningDark" decorative={false} title="warning: " size="sizeIcon30" />;
    case AlertVariants.NEUTRAL:
    default:
      return <InformationIcon iconColor="colorTextIcon" decorative={false} title="information: " size="sizeIcon30" />;
  }
};

const Alert: React.FC<AlertProps> = ({children, onDismiss, variant, role, ...props}) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor={AlertBackgroundColors[variant.toUpperCase()]}
      borderColor={AlertBorderColors[variant.toUpperCase()]}
      borderBottomWidth="borderWidth20"
      borderBottomStyle="solid"
      paddingLeft="space70"
      paddingRight="space70"
      paddingTop="space40"
      paddingBottom="space30"
      role={role != null ? role : AlertRoles[variant.toUpperCase()]}
    >
      <MediaObject as="div">
        <MediaFigure as="div" spacing="space40">
          {renderAlertIcon(variant)}
        </MediaFigure>
        <MediaBody as="div">{children}</MediaBody>
        {onDismiss && typeof onDismiss === 'function' && (
          <MediaFigure align="end" spacing="space70">
            <Button onClick={onDismiss} variant="link" size="reset">
              <CloseIcon iconColor="colorTextIcon" decorative={false} title="dismiss this alert" size="sizeIcon30" />
            </Button>
          </MediaFigure>
        )}
      </MediaObject>
    </Box>
  );
};
Alert.displayName = 'Alert';

if (process.env.NODE_ENV === 'development') {
  Alert.propTypes = {
    children: PropTypes.node.isRequired,
    onDismiss: PropTypes.func,
    role: PropTypes.string,
    variant: PropTypes.oneOf(Object.keys(AlertVariants).map(variant => AlertVariants[variant])).isRequired,
  };
}
export {Alert};
