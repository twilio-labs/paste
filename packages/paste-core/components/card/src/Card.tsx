import { Box, safelySpreadBoxProps } from "@twilio-paste/box";
import type { BoxProps } from "@twilio-paste/box";
import type { PaddingProps } from "@twilio-paste/style-props";
import type { HTMLPasteProps } from "@twilio-paste/types";
import * as React from "react";

/** element identifier from BoxProps for customization */
export interface CardProps extends HTMLPasteProps<"article">, PaddingProps {
  /**
   * Overrides the default element name to apply unique styles with the Customization Provider
   *
   * @default 'CARD'
   * @type {BoxProps['element']}
   * @memberof CardProps
   */
  element?: BoxProps["element"];
}

const Card = React.forwardRef<HTMLElement, CardProps>(
  (
    {
      children,
      element = "CARD",
      padding = "space100",
      paddingBottom,
      paddingLeft,
      paddingRight,
      paddingTop,
      ...props
    },
    ref,
  ) => {
    return (
      <Box
        {...safelySpreadBoxProps(props)}
        ref={ref}
        as="article"
        borderWidth="borderWidth10"
        borderColor="colorBorderWeaker"
        borderStyle="solid"
        borderRadius="borderRadius30"
        element={element}
        padding={padding}
        paddingBottom={paddingBottom}
        paddingLeft={paddingLeft}
        paddingRight={paddingRight}
        paddingTop={paddingTop}
        backgroundColor="colorBackgroundBody"
      >
        {children}
      </Box>
    );
  },
);

Card.displayName = "Card";

export { Card };
