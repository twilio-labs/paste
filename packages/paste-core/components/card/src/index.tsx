export { Card } from "./Card";
export type { CardProps } from "./Card";
