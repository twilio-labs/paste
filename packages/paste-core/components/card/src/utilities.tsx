const paddingProps = ['padding', 'paddingTop', 'paddingRight', 'paddingBottom', 'paddingLeft'];

const errorOnBadProps = (props: {[key: string]: any}, propWhitelist: string[]): void => {
  // Object.keys might be slow in the hot path
  const badProps = Object.keys(props).filter(
    prop => !propWhitelist.includes(prop) && !prop.startsWith('aria-') && !prop.startsWith('data-')
  );

  if (badProps.length > 0) {
    throw new Error(`${badProps.join(', ')} are invalid for the card component`);
  }
};

export {paddingProps, errorOnBadProps};
