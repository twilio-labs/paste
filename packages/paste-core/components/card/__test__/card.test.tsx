import * as React from 'react';
import {Theme} from '@twilio-paste/theme';
import renderer from 'react-test-renderer';
import {Card, CardFooter} from '../src';

describe('Card', () => {
  it('it should render default values', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <Card />
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('it should error with bad propTypes', (): void => {
    expect(() => {
      renderer.create(
        <Theme.Provider>
          <Card error="blowitup" />
        </Theme.Provider>
      );
    }).toThrow();
  });

  it('it should render default values unless overridden by the component', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <Card paddingBottom="space200" />
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('it should render children', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <Card>I AM A JEDI!!!!</Card>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('it should render CardFooter', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider>
          <CardFooter>Just | footer | things</CardFooter>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
