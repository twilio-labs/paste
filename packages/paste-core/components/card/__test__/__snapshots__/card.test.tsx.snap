// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card it should filter out style props that are not allowed 1`] = `
.emotion-1 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-1 *,
.emotion-1 *::after,
.emotion-1 *::before {
  box-sizing: border-box;
}

.emotion-0 {
  box-sizing: border-box;
  border-width: 2px;
  border-color: #e4e7e9;
  border-style: solid;
  border-radius: 4px;
  padding: 1.25rem;
  background-color: #ffffff;
}

<div
  className="emotion-1"
>
  <article
    className="emotion-0"
  />
</div>
`;

exports[`Card it should render children 1`] = `
.emotion-1 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-1 *,
.emotion-1 *::after,
.emotion-1 *::before {
  box-sizing: border-box;
}

.emotion-0 {
  box-sizing: border-box;
  border-width: 2px;
  border-color: #e4e7e9;
  border-style: solid;
  border-radius: 4px;
  padding: 1.25rem;
  background-color: #ffffff;
}

<div
  className="emotion-1"
>
  <article
    className="emotion-0"
  >
    I AM A JEDI!!!!
  </article>
</div>
`;

exports[`Card it should render default values 1`] = `
.emotion-1 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-1 *,
.emotion-1 *::after,
.emotion-1 *::before {
  box-sizing: border-box;
}

.emotion-0 {
  box-sizing: border-box;
  border-width: 2px;
  border-color: #e4e7e9;
  border-style: solid;
  border-radius: 4px;
  padding: 1.25rem;
  background-color: #ffffff;
}

<div
  className="emotion-1"
>
  <article
    className="emotion-0"
  />
</div>
`;

exports[`Card it should render default values unless overridden by the component 1`] = `
.emotion-1 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-1 *,
.emotion-1 *::after,
.emotion-1 *::before {
  box-sizing: border-box;
}

.emotion-0 {
  box-sizing: border-box;
  border-width: 2px;
  border-color: #e4e7e9;
  border-style: solid;
  border-radius: 4px;
  padding: 1.25rem;
  padding-bottom: 4.75rem;
  background-color: #ffffff;
}

<div
  className="emotion-1"
>
  <article
    className="emotion-0"
  />
</div>
`;
